hamt/shardeddir.go

```diff
@@ -272,20 +275,34 @@ func (n UnixFSHAMTShard) Length() int64 {
 		_, pbLink := itr.Next()
 		isValue, err := isValueLink(pbLink, maxPadLen)
 		if err != nil {
-			continue
+			return 0, err
 		}
 		if isValue {
 			total++
 		} else {
 			child, err := n.loadChild(pbLink)
 			if err != nil {
-				continue
+				return 0, err
 			}
-			total += child.Length()
+			cl, err := child.length()
+			if err != nil {
+				return 0, err
+			}
+			total += cl
 		}
 	}
 	n.cachedLength = total
-	return total
+	return total, nil
+}
+
+// Length returns the length of a list, or the number of entries in a map,
+// or -1 if the node is not of list nor map kind.
+func (n UnixFSHAMTShard) Length() int64 {
+	len, err := n.length()
+	if err != nil {
+		return 0
```

Comment:
Don't feel strongly about the number chosen here, as mentioned in https://github.com/ipld/go-ipld-prime/issues/531 we're really in a rough spot here.

This is a behavior change from previously, but if you don't want to return `-1` or a different negative number then `0` is probably more reasonable than what we were doing here before.

Reply:
i'm happy to take suggestions - do you think it makes sense to do `-2` or similar in this case?

Reply:
I'd prefer `0` until we have some kind of agreement on the best way to deal with https://github.com/ipld/go-ipld-prime/issues/531, otherwise this is an outlier.